test/shared/TestConnection.cs

```diff
@@ -108,16 +87,14 @@ public void Dispose()
                 offset += count;
             }
 
-            Assert.Equal(expected, new String(actual, 0, offset));
+            Assert.Equal(expected, new string(actual, 0, offset));
         }
 
         public async Task ReceiveEnd(params string[] lines)
         {
             await Receive(lines);
-            var ch = new char[128];
-            var count = await _reader.ReadAsync(ch, 0, 128).TimeoutAfter(TimeSpan.FromMinutes(1));
-            var text = new string(ch, 0, count);
-            Assert.Equal("", text);
+            _socket.Shutdown(SocketShutdown.Both);
+            await WaitForConnectionClose();
```

Comment:
The reader fails because we shut down the socket:

```
      System.IO.IOException : Unable to read data from the transport connection: A request to send or receive data was disallowed because the socket had already been shut down in that direction with a previous shutdown call.
      ---- System.Net.Sockets.SocketException : A request to send or receive data was disallowed because the socket had already been shut down in that direction with a previous shutdown call
      Stack Trace:
           at System.Net.Sockets.NetworkStream.BeginRead(Byte[] buffer, Int32 offset, Int32 size, AsyncCallback callback, Object state)
           at System.Net.Sockets.NetworkStream.<>c.<ReadAsync>b__58_0(Byte[] bufferArg, Int32 offsetArg, Int32 sizeArg, AsyncCallback callback, Object state)
           at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl[TArg1,TArg2,TArg3](Func`6 beginMethod, Func`2 endFunction, Action`1 endAction, TArg1 arg1, TArg2 arg2, TArg3 arg3, Object state, TaskCreationOptions creationOptions)
           at System.Threading.Tasks.TaskFactory.FromAsync[TArg1,TArg2,TArg3,TResult](Func`6 beginMethod, Func`2 endMethod, TArg1 arg1, TArg2 arg2, TArg3 arg3, Object state)
           at System.Net.Sockets.NetworkStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 size, CancellationToken cancellationToken)
           at System.IO.StreamReader.<ReadBufferAsync>d__102.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.IO.StreamReader.<ReadAsyncInternal>d__69.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at Microsoft.AspNetCore.Testing.TaskExtensions.<TimeoutAfter>d__0`1.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
        D:\src\aspnet\KestrelHttpServer\test\shared\TestConnection.cs(99,0): at Microsoft.AspNetCore.Testing.TestConnection.<ReceiveEnd>d__9.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
        D:\src\aspnet\KestrelHttpServer\test\Microsoft.AspNetCore.Server.Kestrel.FunctionalTests\RequestHeaderLimitsTests.cs(32,0): at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.RequestHeaderLimitsTests.<ServerAcceptsRequestWithHeaderTotalSizeWithinLimit>d__0.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        ----- Inner Stack Trace -----
           at System.Net.Sockets.Socket.BeginReceive(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags, AsyncCallback callback, Object state)
           at System.Net.Sockets.NetworkStream.BeginRead(Byte[] buffer, Int32 offset, Int32 size, AsyncCallback callback, Object state)
```

I'll improve `WaitForConnectionClose()` to print unexpected data in case it fails.